### PR TITLE
Fix implicit points setters

### LINKED ISSUE #713: UniformGrid.points setter not stable - need to deprecate
Hi everyone,

I ran into an issue with the UniformGrid's points.setter method giving me the error:
```
Traceback (most recent call last):
...
...
  File "/Users/Daniel/opt/anaconda3/lib/python3.7/site-packages/pyvista/core/grid.py", line 435, in _from_specs
    self.SetSpacing(xs, ys, zs)

TypeError: SetSpacing argument 1: only size-1 arrays can be converted to Python scalars
```
I use spacings in my point grid that are float and not integer since I need to have my grid in mm coordinates. It seems that the way the spacings are determined  from the point data is not stable in this case. In this case the np.unique call that is supposed to pick one of the spacings determined from the dataset return an array, which then causes the above error. 

Interestingly there is a remark in the code:
TODO: this needs to be tested (unique might return a tuple)
```python
dx, dy, dz = np.unique(dx), np.unique(dy), np.unique(dz)
```
Well, I DID the testing now, is DOES return a tuple :-). In general it does not seem so easy to deal with. I suspect that this is caused by floating point issues. The idea to take the points, reconstruct the grid parameters they have been created with and then recreate the grid seems prone to errors in itself. But I do not have good solution for this. Anyone a good idea? 

-----

**To Reproduce**

```py
import numpy as np
import pyvista as pv

du, dv = (0.2, 0.2)
n_u, n_v = (20, 20)
u, v = du * (np.arange(n_u) + 0.5), dv * (np.arange(n_v) + 0.5)
w = 0.0
pixel_grid_uvw = np.vstack(np.meshgrid(u, v, w))

# remove pixel matrix sub-structure by reshaping
pixel_centers = np.reshape(np.dstack(np.split(pixel_grid_uvw, 3, 0)), (-1, 3))
signal = np.ones(pixel_centers.size)

proj_image = pv.UniformGrid()
proj_image.points = pixel_centers
proj_image['signal'] = signal

```
**System Information:**

```txt
 Date: Wed May 13 12:12:33 2020 CEST

                OS : Darwin
            CPU(s) : 16
           Machine : x86_64
      Architecture : 64bit
               RAM : 32.0 GB
       Environment : Jupyter
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon Pro 5500M OpenGL Engine
       GPU Version : 4.1 ATI-3.8.24

  Python 3.7.7 (default, Mar 26 2020, 10:32:53)  [Clang 4.0.1
  (tags/RELEASE_401/final)]

           pyvista : 0.24.2
               vtk : 8.2.0
             numpy : 1.18.1
           imageio : 2.8.0
           appdirs : 1.4.3
            scooby : 0.5.4
        matplotlib : 3.1.3
             PyQt5 : 5.9.2
           IPython : 7.13.0
             scipy : 1.4.1

  Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for
  Intel(R) 64 architecture applications

```

### PR DESCRIPTION
**Pull request by banesullivan**
### Overview

Resolve #713 

The `UniformGrid` and `RectilinearGrid`'s `points` setter should not be allowed. These changes deprecate those setters to resolve the bug in #713 and address an issue where setting the points in place has no effect.

### Details

- The `UniformGrid` points are implicitly defined and thus we cannot directly set them or directly access/modify them. 
- The `RectilinearGrid` points are similarly not explicitly defined and must be set along each axial direction individually.

Now when a user tries to set the points for either of these types, they will see a detailed error message:

```py
>>> from pyvista import examples
>>> import numpy as np
>>> grid = examples.load_uniform()
>>> grid.points = np.random.rand(10, 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bane/Software/pyvista/pyvista/pyvista/core/grid.py", line 370, in points
    raise AttributeError("The points cannot be set. The points of "
AttributeError: The points cannot be set. The points of `UniformGrid`/`vtkImageData` are implicitly defined by the `origin`, `spacing`, and `dimensions` of the grid.
>>>
```

Unfortunately, there still remains a behavior where the user might update the points in place and no error is raised or no effect takes place:

```py
>>> import numpy as np
>>> from pyvista import examples

>>> a = examples.load_rectilinear()
>>> b = a.copy()
>>> a.points[:,0] *= 10000000
>>> np.allclose(a.points, b.points)
True
```

This is because the numpy array that was fetched from the `points` getter is modified in place and not passed back to the setter. This could still confuse users, but I think we have to leave it as is because a user may want to modify that array of points in their own workflow. To mitigate this, I updated the language in the docs for those classes' `points` getter to highlight that they return a COPY of the points.

### CODE REVIEW

File: pyvista/core/grid.py [hunk 1]
```diff
@@ -159,33 +159,35 @@ def _from_arrays(self, x, y, z):
...
-        # Set the vtk coordinates
-        self._from_arrays(x, y, z)
-        #self._point_ref = points
-        self.Modified()
+        """Points must be set along each axial direction.
+
+        Please set the point coordinates with the ``x``, ``y``, and ``z``
+        setters.
+
+        This setter overrides the base class' setter to ensure a user does not
```
Comment (akaszynski) [inline pyvista/core/grid.py:183]:
```suggestion
        This setter overrides the base class's setter to ensure a user does not
```
I had to look this up: https://english.stackexchange.com/a/44085

File: pyvista/core/grid.py [hunk 2]
```diff
@@ -359,23 +361,16 @@ def points(self):
...
-        # TODO: this needs to be tested (unique might return a tuple)
-        dx, dy, dz = np.unique(dx), np.unique(dy), np.unique(dz)
-        ox, oy, oz = np.min(x), np.min(y), np.min(z)
-        # Build the vtk object
-        self._from_specs((nx,ny,nz), (dx,dy,dz), (ox,oy,oz))
-        #self._point_ref = points
-        self.Modified()
+        """Points cannot be set.
+
+        This setter overrides the base class' setter to ensure a user does not
```
Comment (akaszynski) [inline pyvista/core/grid.py:366]:
```suggestion
        This setter overrides the base class's setter to ensure a user does not
```
Same here.

File: tests/test_grid.py [hunk 3]
```diff
@@ -442,14 +442,20 @@ def test_save_uniform(extension, binary, tmpdir):
...
-    grid.points = np.array([x_surf, y_surf, z_surf]).transpose()
+    with pytest.raises(AttributeError):
+        grid.points = points
+    grid.origin = (0.0, 0.0, 0.0)
+    grid.dimensions = (3, 3, 1)
+    grid.spacing = (1, 1, 1)
     assert grid.n_points == 9
     assert grid.n_cells == 4
+    assert np.allclose(grid.points, points)
     del grid
```
Comment (akaszynski) [inline tests/test_grid.py:458]: Curious, why are we deleting grid here?  If we want to check for garbage collection we might want a separate test.

File: tests/test_grid.py [hunk 4]
```diff
@@ -459,19 +465,26 @@ def test_grid_points():
...
-    assert grid.origin == [0., 0., 0.]
+    grid.dimensions = [2, 2, 2]
+    grid.spacing = [1, 1, 1]
+    grid.origin = [0., 0., 0.]
     assert np.allclose(np.unique(grid.points, axis=0), np.unique(points, axis=0))
     opts = np.c_[grid.x, grid.y, grid.z]
     assert np.allclose(np.unique(opts, axis=0), np.unique(points, axis=0))
+
     # Now test rectilinear grid
     del grid
```
Comment (akaszynski) [inline tests/test_grid.py:476]: Same here.  We might consider having separate tests as unit tests are generally more isolated in scope.

File: tests/test_grid.py (hunk 3 above)
Comment (banesullivan) [reply tests/test_grid.py:458]: I had thought the same thing when updating this test... I’m not sure why. I will remove this

Garbage collection should be all good as far as these copies of the points go from what I can tell

File: pyvista/core/grid.py (hunk 1 above)
Comment (banesullivan) [reply pyvista/core/grid.py:183]: Ahhh thanks! My grade school English teacher would be disappointed in me

**Comment by akaszynski** [comment]
@banesullivan, I'm going to have a RC by Wednesday.  Let's have this merged before then.

**Comment by banesullivan** [comment]
Ready

**Pull request merged**